--- kernel-2.6.spec.dist	2006-03-31 20:19:41.000000000 -0500
+++ kernel-2.6.spec	2006-03-31 20:26:56.000000000 -0500
@@ -288,6 +288,8 @@
 
 # Tux http accelerator.
 Patch910: linux-2.6-tux.patch
+# A temporary patch for Frysk project
+Patch920: linux-2.6-strace-clone-exec.patch
 
 #
 # Patches 1000 to 5000 are reserved for bugfixes to drivers and filesystems
@@ -869,6 +871,8 @@
 
 # Tux
 %patch910 -p1
+# Frysk
+%patch920 -p1
 
 #
 # Patches 1000 to 5000 are reserved for bugfixes to drivers and filesystems
